lib/services/s3.js

```diff
@@ -338,22 +428,172 @@ AWS.util.update(AWS.S3.prototype, {
+  },
+
+  /**
+   * If region was not obtained synchronously, then send async request
+   * to get bucket region for errors resulting from wrong region.
+   *
+   * @api private
+   */
+   requestBucketRegion: function requestBucketRegion(resp, done) {
+    var error = resp.error;
+    var req = resp.request;
+    var bucket = req.params.Bucket || null;
+
+    if (!error || !bucket || error.region || req.operation === 'listObjects' ||
+        (AWS.util.isNode() && req.operation === 'headBucket') ||
+        (error.statusCode === 400 && req.operation !== 'headObject') ||
+        regionRedirectErrorCodes.indexOf(error.code) === -1) {
+      return done();
+    }
+
+    var regionReq = req.service.listObjects({Bucket: bucket});
+    regionReq._requestRegionForBucket = bucket;
+    regionReq.send(function() {
+      var region = req.service.bucketRegionCache[bucket] || null;
+      error.region = region;
+      done();
+    });
+   },
+
+   /**
+   * For browser only. If NetworkingError received, will attempt to obtain
+   * the bucket region.
+   *
+   * @api private
+   */
+   reqRegionForNetworkingError: function reqRegionForNetworkingError(resp, done) {
+    if (!AWS.util.isBrowser()) {
+      return done();
+    }
+    var error = resp.error;
+    var request = resp.request;
+    var bucket = request.params.Bucket;
+    if (!error || error.code !== 'NetworkingError' || !bucket ||
+        request.httpRequest.region === 'us-east-1') {
+      return done();
+    }
+    var service = request.service;
+    var bucketRegionCache = service.bucketRegionCache;
+    var cachedRegion = bucketRegionCache[bucket] || null;
+
+    if (cachedRegion && cachedRegion !== request.httpRequest.region) {
+      service.updateReqBucketRegion(request, cachedRegion);
+      done();
+    } else if (!service.dnsCompatibleBucketName(bucket)) {
+      service.updateReqBucketRegion(request, 'us-east-1');
+      if (bucketRegionCache[bucket] !== 'us-east-1') {
+        bucketRegionCache[bucket] = 'us-east-1';
+      }
+      done();
+    } else if (request.httpRequest.virtualHostedBucket) {
+      var getRegionReq = service.listObjects({Bucket: bucket});
+      service.updateReqBucketRegion(getRegionReq, 'us-east-1');
+      getRegionReq._requestRegionForBucket = bucket;
+
+      getRegionReq.send(function() {
+        var region = service.bucketRegionCache[bucket] || null;
+        if (region && region !== request.httpRequest.region) {
+          service.updateReqBucketRegion(request, region);
+        }
+        done();
       });
+    } else {
+      // DNS-compatible path-style
+      // (s3ForcePathStyle or bucket name with dot over https)
+      // Cannot obtain region information for this case
+      done();
+    }
+   },
+
+  /**
+   * Cache for bucket region.
+   *
+   * @api private
+   */
+   bucketRegionCache: {},
+
+  /**
+   * Clears bucket region cache.
+   *
+   * @api private
+   */
+   clearBucketRegionCache: function(buckets) {
```

Comment:
The SDK doesn't actually ever call `clearBucketRegionCache` without passing in bucket, except for in the unit tests where almost all of the tests store at most 1 bucket, so it shouldn't cause any performance issues in the SDK. The reason I decided to delete properties rather than replace the cache with a whole new object was that if customers wanted to call the method themselves, and they had assigned the `bucketRegionCache` to a variable (so they could avoid typing long chains such as `response.request.service.bucketRegionCache` multiple times), then `clearBucketRegionCache` would still behave the way they expect, and the `bucketRegionCache` they stored in their variable would be cleared and still valid. If I just replaced `s3.bucketRegionCache` with a new object, not only would that not clear the cache that they have stored in their variable, any modifications they make to the cache stored in their variable would unexpectedly no longer be accessible to the SDK. So the decision was mainly rooted in a more intuitive customer experience, but I don't see it making too big a difference either way, since the method is marked as private and I don't think there are many scenarios in which users would want to clear the entire cache instead of just a select number of buckets.